tests/subsurfaces.cpp

```diff
@@ -0,0 +1,215 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Authored by: William Wold <williamwold@canonical.com>
+ */
+
+#include "helpers.h"
+#include "in_process_server.h"
+#include "xdg_shell_v6.h"
+
+#include <gmock/gmock.h>
+
+#include <vector>
+
+using namespace testing;
+
+class SubsurfaceTest : public wlcs::StartedInProcessServer
+{
+public:
+    static int const surface_width = 200, surface_height = 300;
+    static int const subsurface_width = 50, subsurface_height = 50;
+    static int const surface_x = 20, surface_y = 30;
+
+    SubsurfaceTest()
+    {
+        the_server().move_surface_to(main_surface, surface_x, surface_y);
+        client.roundtrip();
+    }
+
+    void move_subsurface_to(int x, int y)
+    {
+        wl_subsurface_set_position(subsurface, x, y);
+        wl_surface_commit(subsurface);
+        wl_surface_commit(main_surface);
+        client.roundtrip();
+    }
+
+    wlcs::Client client{the_server()};
+    wlcs::Surface main_surface{client.create_visible_surface(surface_width, surface_height)};
+    wlcs::Subsurface subsurface{wlcs::Subsurface::create_visible(main_surface, 0, 0, subsurface_width, subsurface_height)};
+};
+
+TEST_F(SubsurfaceTest, subsurface_can_be_created)
+{
+    EXPECT_THAT(&subsurface.parent(), Eq(&main_surface));
+}
+
+TEST_F(SubsurfaceTest, gets_pointer_input)
+{
+    int const pointer_x = surface_x + 10, pointer_y = surface_y + 5;
+
+    auto pointer = the_server().create_pointer();
+    pointer.move_to(pointer_x, pointer_y);
+    client.roundtrip();
+
+    EXPECT_THAT(client.focused_window(), Ne((wl_surface*)main_surface)) << "input fell through to main surface";
+    EXPECT_THAT(client.focused_window(), Eq((wl_surface*)subsurface));
+    EXPECT_THAT(client.pointer_position(),
+                Eq(std::make_pair(
+                    wl_fixed_from_int(pointer_x - surface_x),
+                    wl_fixed_from_int(pointer_y - surface_y))));
+}
+
+TEST_F(SubsurfaceTest, pointer_input_offset)
+{
+    int const pointer_x = surface_x + 13, pointer_y = surface_y + 24;
+    int const subsurface_x = 8, subsurface_y = 17;
+
+    wl_subsurface_set_position(subsurface, subsurface_x, subsurface_y);
+    wl_surface_commit(subsurface);
+    wl_surface_commit(main_surface);
+    client.roundtrip();
+
+    auto pointer = the_server().create_pointer();
+    pointer.move_to(pointer_x, pointer_y);
+    client.roundtrip();
+
+    EXPECT_THAT(client.focused_window(), Ne((wl_surface*)main_surface)) << "input fell through to main surface";
+    EXPECT_THAT(client.focused_window(), Eq((wl_surface*)subsurface));
+    EXPECT_THAT(client.pointer_position(),
+                Eq(std::make_pair(
+                    wl_fixed_from_int(pointer_x - surface_x - subsurface_x),
+                    wl_fixed_from_int(pointer_y - surface_y - subsurface_y))));
+}
+
+// appears to work in normal Mir, but breaks in WLCS
```

Comment:
Appears OK in wlcs

Reply:
Ah yes, I fixed that in the wm a while back.